File: core/src/validation.ts
import { isIP } from 'node:net';
import { SendSoonErrorCode, createError, type SendSoonError } from './errors.js';
import {
  MAX_MARKITDOWN_FILE_BYTES,
  type MarkitdownConvertRequest,
} from './types/markitdown.js';
import {
  MAX_BODY_LENGTH,
  MAX_FROM_ALIAS_LENGTH,
  MAX_SUBJECT_LENGTH,
  type SendRequest,
} from './types/send-request.js';

const SUPPORTED_MARKITDOWN_EXTENSIONS = new Set([
  '.pdf', '.pptx', '.docx', '.xlsx', '.xls', '.jpg', '.jpeg', '.png',
  '.gif', '.bmp', '.tiff', '.mp3', '.wav', '.m4a', '.html', '.htm',
  '.csv', '.json', '.xml', '.zip', '.epub', '.txt', '.md',
]);

function ipv4Bytes(address: string): number[] {
  return address.split('.').map(Number);
}

function isPublicIpv4(address: string): boolean {
  const [a, b, c] = ipv4Bytes(address);
  return !(
    a === 0 ||
    a === 10 ||
    a === 127 ||
    (a === 100 && b >= 64 && b <= 127) ||
    (a === 169 && b === 254) ||
    (a === 172 && b >= 16 && b <= 31) ||
    (a === 192 && b === 0 && c === 0) ||
    (a === 192 && b === 0 && c === 2) ||
    (a === 192 && b === 88 && c === 99) ||
    (a === 192 && b === 168) ||
    (a === 198 && (b === 18 || b === 19)) ||
    (a === 198 && b === 51 && c === 100) ||
    (a === 203 && b === 0 && c === 113) ||
    a >= 224
  );
}

function expandIpv6(address: string): number[] | null {
  const normalized = address.toLowerCase().split('%')[0];
  const [leftRaw, rightRaw, ...extra] = normalized.split('::');
  if (extra.length > 0) {
    return null;
  }

  const parseSide = (side: string): number[] | null => {
    if (!side) return [];
    const result: number[] = [];
    for (const part of side.split(':')) {
      if (part.includes('.')) {
        if (isIP(part) !== 4) return null;
        const bytes = ipv4Bytes(part);
        result.push(bytes[0] * 256 + bytes[1], bytes[2] * 256 + bytes[3]);
      } else if (/^[0-9a-f]{1,4}$/.test(part)) {
        result.push(Number.parseInt(part, 16));
      } else {
        return null;
      }
    }
    return result;
  };

  const left = parseSide(leftRaw);
  const right = parseSide(rightRaw ?? '');
  if (!left || !right) return null;

  if (rightRaw === undefined) {
    return left.length === 8 ? left : null;
  }

  const missing = 8 - left.length - right.length;
  return missing >= 1 ? [...left, ...Array<number>(missing).fill(0), ...right] : null;
}

function isPublicIpv6(address: string): boolean {
  const groups = expandIpv6(address);
  if (!groups) return false;

  const [first, second] = groups;
  const allZero = groups.every((group) => group === 0);
  const loopback = groups.slice(0, 7).every((group) => group === 0) && groups[7] === 1;
  const mappedIpv4 = groups.slice(0, 5).every((group) => group === 0) && groups[5] === 0xffff;
  if (mappedIpv4) {
    const embedded = `${groups[6] >> 8}.${groups[6] & 255}.${groups[7] >> 8}.${groups[7] & 255}`;
    return isPublicIpv4(embedded);
  }

  if ((first & 0xe000) !== 0x2000) return false;

  return !(
    allZero ||
    loopback ||
    (first & 0xfe00) === 0xfc00 ||
    (first & 0xffc0) === 0xfe80 ||
    (first & 0xff00) === 0xff00 ||
    (first === 0x2001 && second === 0x0db8) ||
    (first === 0x0100 && second === 0)
  );
}

export function validatePublicIp(ip: string): SendSoonError | null {
  const version = isIP(ip);
  if (version === 0) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'ip must be a valid IPv4 or IPv6 address.');
  }

  const isPublic = version === 4 ? isPublicIpv4(ip) : isPublicIpv6(ip);
  return isPublic
    ? null
    : createError(
        SendSoonErrorCode.INVALID_INPUT,
        'ip must be public; private, reserved, loopback, link-local, and multicast addresses are not supported.',
      );
}

export function validateSendRequest(request: SendRequest): SendSoonError | null {
  const email = request.to.trim();
  const at = email.lastIndexOf('@');
  const local = at > 0 ? email.slice(0, at) : '';
  const domain = at > 0 ? email.slice(at + 1) : '';
  const validDomain = domain.length <= 253 && domain.split('.').length >= 2 &&
    domain.split('.').every((label) =>
      label.length > 0 && label.length <= 63 &&
      /^[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?$/.test(label),
    );
  if (
    email.length > 254 || local.length > 64 || !local ||
    local.startsWith('.') || local.endsWith('.') || local.includes('..') ||
    /[\s@]/.test(local) || !validDomain
  ) {
    return createError(SendSoonErrorCode.INVALID_RECIPIENT);
  }
  if (!request.subject.trim()) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'Subject is required and cannot be empty.');
  }
  if (request.subject.length > MAX_SUBJECT_LENGTH) {
    return createError(SendSoonErrorCode.INVALID_INPUT, `Subject must be at most ${MAX_SUBJECT_LENGTH} characters.`);
  }
  if (!request.body.trim()) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'Body is required and cannot be empty.');
  }
  if (Buffer.byteLength(request.body, 'utf8') > MAX_BODY_LENGTH) {
    return createError(SendSoonErrorCode.PAYLOAD_TOO_LARGE);
  }
  if (request.from_alias && request.from_alias.length > MAX_FROM_ALIAS_LENGTH) {
    return createError(SendSoonErrorCode.INVALID_INPUT, `from_alias must be at most ${MAX_FROM_ALIAS_LENGTH} characters.`);
  }
  if (request.idempotency_key !== undefined) {
    const key = request.idempotency_key.trim();
    if (!key || key.length > 128 || !/^[A-Za-z0-9._:-]+$/.test(key)) {
      return createError(
        SendSoonErrorCode.INVALID_INPUT,
        'idempotency_key must be 1-128 characters using letters, numbers, dot, underscore, colon, or hyphen.',
      );
    }
  }
  return null;
}

function normalizeBase64(value: string): string | null {
  const normalized = value.replace(/\s/g, '');
  if (!normalized || /[^A-Za-z0-9+/=]/.test(normalized) || /=/.test(normalized.slice(0, -2))) {
    return null;
  }
  if (normalized.length % 4 === 1 || !/^[A-Za-z0-9+/]*={0,2}$/.test(normalized)) {
    return null;
  }
  const decoded = Buffer.from(normalized, 'base64');
  const canonical = decoded.toString('base64').replace(/=+$/, '');
  return canonical === normalized.replace(/=+$/, '') ? normalized : null;
}

export function decodedBase64ByteLength(value: string): number | null {
  const normalized = normalizeBase64(value);
  return normalized === null ? null : Buffer.from(normalized, 'base64').byteLength;
}

export function validateMarkitdownRequest(request: MarkitdownConvertRequest): SendSoonError | null {
  const filename = request.filename.trim();
  if (!filename) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'filename is required and cannot be empty.');
  }
  if (filename.includes('/') || filename.includes('\\') || filename.includes('\0')) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'filename must be a base name without path separators.');
  }
  const dot = filename.lastIndexOf('.');
  const extension = dot >= 0 ? filename.slice(dot).toLowerCase() : '';
  if (!SUPPORTED_MARKITDOWN_EXTENSIONS.has(extension)) {
    return createError(SendSoonErrorCode.INVALID_INPUT, `Unsupported file extension: ${extension || '(none)'}.`);
  }

  const byteLength = decodedBase64ByteLength(request.content_base64);
  if (byteLength === null) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'content_base64 must contain valid Base64 data.');
  }
  if (byteLength === 0) {
    return createError(SendSoonErrorCode.INVALID_INPUT, 'The file cannot be empty.');
  }
  if (byteLength > MAX_MARKITDOWN_FILE_BYTES) {
    return createError(SendSoonErrorCode.PAYLOAD_TOO_LARGE);
  }
  return null;
}

export function validateBaseUrl(baseUrl: string): SendSoonError | null {
  try {
    const url = new URL(baseUrl);
    const local = ['localhost', '127.0.0.1', '::1'].includes(url.hostname);
    if (url.protocol !== 'https:' && !(local && url.protocol === 'http:')) {
      return createError(SendSoonErrorCode.INVALID_CONFIG, 'SENDSOON_API_BASE_URL must use HTTPS (HTTP is allowed only for localhost).');
    }
    if (url.username || url.password) {
      return createError(SendSoonErrorCode.INVALID_CONFIG, 'SENDSOON_API_BASE_URL must not include credentials.');
    }
    if (url.search || url.hash) {
      return createError(SendSoonErrorCode.INVALID_CONFIG, 'SENDSOON_API_BASE_URL must not include a query string or fragment.');
    }
    return null;
  } catch {
    return createError(SendSoonErrorCode.INVALID_CONFIG, 'SENDSOON_API_BASE_URL must be a valid URL.');
  }
}

export const MARKITDOWN_EXTENSIONS = [...SUPPORTED_MARKITDOWN_EXTENSIONS];

File: pypi/src/sendsoon_mcp/validation.py
"""Request validation helpers (aligned with npm/src/core)."""

from __future__ import annotations

import base64
import binascii
import ipaddress
import re
from typing import Any
from urllib.parse import urlparse

from .errors import SendSoonError, create_error

MAX_BODY_LENGTH = 512_000
MAX_SUBJECT_LENGTH = 998
MAX_MARKITDOWN_FILE_BYTES = 10 * 1024 * 1024

SUPPORTED_MARKITDOWN_EXTENSIONS = frozenset(
    {
        ".pdf",
        ".docx",
        ".pptx",
        ".xlsx",
        ".xls",
        ".html",
        ".htm",
        ".txt",
        ".md",
    }
)

_DOMAIN_LABEL_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?$")
_IDEMPOTENCY_KEY_RE = re.compile(r"^[A-Za-z0-9._:-]+$")
_BASE64_CHAR_RE = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")


def validate_public_ip(ip: str) -> SendSoonError | None:
    try:
        address = ipaddress.ip_address(ip)
    except ValueError:
        return create_error("INVALID_INPUT", "ip must be a valid IPv4 or IPv6 address.")

    if address.is_global:
        return None

    return create_error(
        "INVALID_INPUT",
        "ip must be public; private, reserved, loopback, link-local, "
        "and multicast addresses are not supported.",
    )


def validate_send_request(request: dict[str, Any]) -> SendSoonError | None:
    email = str(request.get("to", "")).strip()
    at = email.rfind("@")
    local = email[:at] if at > 0 else ""
    domain = email[at + 1 :] if at > 0 else ""
    labels = domain.split(".")
    valid_domain = (
        len(domain) <= 253
        and len(labels) >= 2
        and all(0 < len(label) <= 63 and _DOMAIN_LABEL_RE.match(label) for label in labels)
    )
    if (
        len(email) > 254
        or len(local) > 64
        or not local
        or local.startswith(".")
        or local.endswith(".")
        or ".." in local
        or " " in local
        or "@" in local
        or not valid_domain
    ):
        return create_error("INVALID_RECIPIENT")

    subject = str(request.get("subject", ""))
    if not subject.strip():
        return create_error("INVALID_INPUT", "Subject is required and cannot be empty.")
    if len(subject) > MAX_SUBJECT_LENGTH:
        return create_error(
            "INVALID_INPUT",
            f"Subject must be at most {MAX_SUBJECT_LENGTH} characters.",
        )

    body = str(request.get("body", ""))
    if not body.strip():
        return create_error("INVALID_INPUT", "Body is required and cannot be empty.")
    if len(body.encode("utf-8")) > MAX_BODY_LENGTH:
        return create_error("PAYLOAD_TOO_LARGE")

    if "idempotency_key" in request and request["idempotency_key"] is not None:
        key = str(request["idempotency_key"]).strip()
        if not key or len(key) > 128 or not _IDEMPOTENCY_KEY_RE.match(key):
            return create_error(
                "INVALID_INPUT",
                "idempotency_key must be 1-128 characters using letters, numbers, "
                "dot, underscore, colon, or hyphen.",
            )
    return None


def _normalize_base64(value: str) -> str | None:
    normalized = re.sub(r"\s+", "", value)
    if not normalized or re.search(r"[^A-Za-z0-9+/=]", normalized):
        return None
    if "=" in normalized[:-2]:
        return None
    if len(normalized) % 4 == 1 or not _BASE64_CHAR_RE.match(normalized):
        return None
    try:
        decoded = base64.b64decode(normalized, validate=True)
    except (binascii.Error, ValueError):
        return None
    canonical = base64.b64encode(decoded).decode("ascii").rstrip("=")
    return normalized if canonical == normalized.rstrip("=") else None


def decoded_base64_byte_length(value: str) -> int | None:
    normalized = _normalize_base64(value)
    if normalized is None:
        return None
    return len(base64.b64decode(normalized, validate=True))


MARKITDOWN_EXTENSIONS = tuple(sorted(SUPPORTED_MARKITDOWN_EXTENSIONS))


def validate_markitdown_filename(filename: str) -> SendSoonError | None:
    trimmed = filename.strip()
    if not trimmed:
        return create_error("INVALID_INPUT", "filename is required and cannot be empty.")
    if "/" in trimmed or "\\" in trimmed or "\0" in trimmed:
        return create_error(
            "INVALID_INPUT",
            "filename must be a base name without path separators.",
        )

    dot = trimmed.rfind(".")
    extension = trimmed[dot:].lower() if dot >= 0 else ""
    if extension not in SUPPORTED_MARKITDOWN_EXTENSIONS:
        return create_error(
            "INVALID_INPUT",
            f"Unsupported file extension: {extension or '(none)'}.",
        )
    return None


def validate_markitdown_request(request: dict[str, Any]) -> SendSoonError | None:
    filename_error = validate_markitdown_filename(str(request.get("filename", "")))
    if filename_error:
        return filename_error

    content_base64 = str(request.get("content_base64", ""))
    byte_length = decoded_base64_byte_length(content_base64)
    if byte_length is None:
        return create_error("INVALID_INPUT", "content_base64 must contain valid Base64 data.")
    if byte_length == 0:
        return create_error("INVALID_INPUT", "The file cannot be empty.")
    if byte_length > MAX_MARKITDOWN_FILE_BYTES:
        return create_error("PAYLOAD_TOO_LARGE")
    return None


def validate_base_url(base_url: str) -> SendSoonError | None:
    try:
        parsed = urlparse(base_url)
    except ValueError:
        return create_error("INVALID_CONFIG", "SENDSOON_API_BASE_URL must be a valid URL.")

    if not parsed.scheme or not parsed.netloc:
        return create_error("INVALID_CONFIG", "SENDSOON_API_BASE_URL must be a valid URL.")

    host = parsed.hostname or ""
    local = host in {"localhost", "127.0.0.1", "::1"}
    if parsed.scheme != "https" and not (local and parsed.scheme == "http"):
        return create_error(
            "INVALID_CONFIG",
            "SENDSOON_API_BASE_URL must use HTTPS (HTTP is allowed only for localhost).",
        )
    if parsed.username is not None or parsed.password is not None:
        return create_error(
            "INVALID_CONFIG",
            "SENDSOON_API_BASE_URL must not include credentials.",
        )
    if parsed.query or parsed.fragment:
        return create_error(
            "INVALID_CONFIG",
            "SENDSOON_API_BASE_URL must not include a query string or fragment.",
        )
    return None
